server/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java

```diff
@@ -0,0 +1,284 @@
+         * - bring node2 back
+         * -> nodes are fully in-sync and has different to initial history uuid
+         */
+
+        // initial set up
+        final String indexName = "index42";
+        final String master = internalCluster().startMasterOnlyNode();
+        String node1 = internalCluster().startNode();
+        createIndex(indexName);
+        final int numDocs = indexDocs(indexName, "foo", "bar");
+        final IndexSettings indexSettings = getIndexSettings(indexName, node1);
+        final String primaryNodeId = getNodeIdByName(node1);
+        final Set<String> allocationIds = getAllocationIds(indexName);
+        final ShardId shardId = new ShardId(resolveIndex(indexName), 0);
+        final Path indexPath = getIndexPath(node1, shardId);
+        assertThat(allocationIds, hasSize(1));
+        final Set<String> historyUUIDs = historyUUIDs(node1, indexName);
+        String node2 = internalCluster().startNode();
+        ensureGreen(indexName);
+        internalCluster().assertSameDocIdsOnShards();
+        // initial set up is done
+
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(node1));
+
+        // index more docs to node2 that marks node1 as stale
+        int numExtraDocs = indexDocs(indexName, "foo", "bar2");
+        assertHitCount(client(node2).prepareSearch(indexName).setQuery(matchAllQuery()).get(), numDocs + numExtraDocs);
+
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(node2));
+
+        // create fake corrupted marker on node1
+        putFakeCorruptionMarker(indexSettings, shardId, indexPath);
+
+        // thanks to master node1 is out of sync
+        node1 = internalCluster().startNode();
+
+        // there is only _stale_ primary
+        checkNoValidShardCopy(indexName, shardId);
+
+        // allocate stale primary
+        client(node1).admin().cluster().prepareReroute()
+            .add(new AllocateStalePrimaryAllocationCommand(indexName, 0, primaryNodeId, true))
+            .get();
+
+        // allocation fails due to corruption marker
+        assertBusy(() -> {
+            final ClusterState state = client().admin().cluster().prepareState().get().getState();
+            final ShardRouting shardRouting = state.routingTable().index(indexName).shard(shardId.id()).primaryShard();
+            assertThat(shardRouting.state(), equalTo(ShardRoutingState.UNASSIGNED));
+            assertThat(shardRouting.unassignedInfo().getReason(), equalTo(UnassignedInfo.Reason.ALLOCATION_FAILED));
+        });
+
+        try(Store store = new Store(shardId, indexSettings, new SimpleFSDirectory(indexPath), new DummyShardLock(shardId))) {
+            store.removeCorruptionMarker();
+        }
+
+        // open index; no any reasons to wait 30 sec as no any health shard for that
+        client(node1).admin().indices().prepareOpen(indexName).setTimeout(TimeValue.timeValueSeconds(5)).get();
```

Comment:
why are we opening the index here? it is not closed?

Reply:
Shard is failed due to corruption marked; after corrupted marker is removed I reopen index instead of restart entire node to check that it could not proceed further due to fake allocation id